src/expressions/FlworExpression.ts

```diff
@@ -0,0 +1,239 @@
+	): IAsyncIterator<UpdatingExpressionResult> {
+		return this.doFlworExpressionUpdating(
+			dynamicContext,
+			createSingleValueIterator(dynamicContext),
+			executionParameters
+		);
+	}
+
+	public performFunctionalEvaluation(
+		_dynamicContext: DynamicContext,
+		_executionParameters: ExecutionParameters,
+		_sequenceCallbacks: SequenceCallbacks
+	) {
+		throw new Error('TODO better modeling');
+
+		return sequenceFactory.empty();
+	}
+
+	public doFlworExpressionUpdating(
+		dynamicContext: DynamicContext,
+		dynamicContextIterator: IAsyncIterator<DynamicContext>,
+		executionParameters: ExecutionParameters
+	): IAsyncIterator<UpdatingExpressionResult> {
+		let updateList = [];
+		const sequence = this.doFlworExpression(
+			dynamicContext,
+			dynamicContextIterator,
+			executionParameters,
+			(dynamicContextIterator) => {
+				if (this._returnExpression instanceof FlworExpression) {
+					// We are in a FLWOR, the return is also a FLWOR, keep piping dynamiccontext generators
+					const updateListAndValue = this._returnExpression.doFlworExpressionUpdating(
+						dynamicContext,
+						dynamicContextIterator,
+						executionParameters
+					);
+
+					return separateXDMValueFromUpdatingExpressionResult(
+						updateListAndValue,
+						(pendingUpdates) => (updateList = pendingUpdates)
+					);
+				}
+				// We are transitioning to a non-FLWOR expression. Also apply all looping here
+				let currentReturnValueGenerator: IAsyncIterator<Value> = null;
+				return sequenceFactory.create({
+					next: (_hint) => {
+						while (true) {
+							if (!currentReturnValueGenerator) {
+								const currentDynamicContext = dynamicContextIterator.next(
+									IterationHint.NONE
+								);
+
+								if (currentDynamicContext.done) {
+									return DONE_TOKEN;
+								}
+
+								const updateListAndValue = this._returnExpression.evaluateWithUpdateList(
+									currentDynamicContext.value,
+									executionParameters
+								);
+								currentReturnValueGenerator = separateXDMValueFromUpdatingExpressionResult(
+									updateListAndValue,
+									(pendingUpdates) =>
+										(updateList = mergeUpdates(updateList, pendingUpdates))
+								).value;
+							}
+
+							const nextValue = currentReturnValueGenerator.next(IterationHint.NONE);
+							if (nextValue.done) {
+								currentReturnValueGenerator = null;
+								continue;
+							}
+							return nextValue;
+						}
+					},
+				});
+			}
+		);
+
+		let done = false;
+		return {
+			next: (_hint: IterationHint) => {
+				if (done) {
+					return DONE_TOKEN;
+				}
+				// Ensure we fully exhaust the inner expression so that the pending update list is
+				// filled
+				const allValues = sequence.tryGetAllValues();
+				if (!allValues.ready) {
+					return notReady(allValues.promise);
+				}
+				done = true;
+				return ready(new UpdatingExpressionResult(allValues.value, updateList));
+			},
+		};
+	}
+
+	public doFlworExpressionA(
```

Comment:
Let's rename this to `doFlworExpressionInternal` or something

Reply:
This can also be private I think

Reply:
Done